src/executor.rs

```diff
@@ -25,161 +30,90 @@ pub struct Task {
+        let (stop_sender, mut stop_receiver) = oneshot::channel();
+        let (line_sender, line_receiver) = crossbeam::channel::unbounded();
         command
+            .stdin(Stdio::null())
             .stderr(Stdio::piped())
-            .stdout(if with_stdout { Stdio::piped() } else { Stdio::null() });
+            .stdout(if with_stdout {
+                Stdio::piped()
+            } else {
+                Stdio::null()
+            });
+
         let thread = thread::spawn(move || {
-            loop {
-                select! {
-                    recv(task_receiver) -> task => {
-                        let task = match task {
-                            Ok(task) => task,
-                            _ => { break; }
-                        };
-                        debug!("starting task {:?}", task);
-                        command.env(
-                            "RUST_BACKTRACE",
-                            if task.backtrace { "1" } else { "0" },
-                        );
-                        let child = command.spawn();
-                        let mut child = match child {
-                            Ok(child) => child,
-                            Err(e) => {
-                                if let Err(e) = line_sender.send(CommandExecInfo::Error(
-                                    format!("command launch failed: {}", e)
-                                )) {
-                                    debug!("error when sending launch error: {}", e);
-                                    break;
-                                }
-                                continue;
-                            }
-                        };
-                        let stderr = match child.stderr.take() {
-                            Some(stderr) => stderr,
-                            None => {
-                                if let Err(e) = line_sender.send(CommandExecInfo::Error(
-                                    "taking stderr failed".to_string()
-                                )) {
-                                    debug!("error when sending stderr error: {}", e);
-                                    break;
-                                }
-                                continue;
+            // start a runtime to manage the executor
+            let rt = tokio::runtime::Builder::new_current_thread()
+                .enable_io()
+                .build()
+                .unwrap();
+
+            rt.block_on(async move {
+                // Handle to the current task
+                let mut current_task: Option<tokio::task::JoinHandle<_>> = None;
+
+                loop {
+                    tokio::select! {
+                        _ = &mut stop_receiver => break,
+
+                        // wait for the next task
+                        task = task_receiver.recv() => {
+                            if let Some(old) = current_task.take() {
+                                old.abort();
```

Comment:
where is define this ? `Task` is define in this file but abort is nowhere to be found

Reply:
This is a tokio task (`abort` is in `tokio::task::JoinHandle`)

Reply:
It's unclear why you use a task and not stock directly the Command and use `kill()`. I unsure this correctly clean the child process.

Reply:
It will, due to the call to `kill_on_drop` on the command builder (https://github.com/Canop/bacon/blob/fd714b62e4760abac8947c7a6fc5886fc5cd0f72/src/executor.rs#L164).